### ISSUE TITLE
Unable to use AWS ECR as Private Registry for System Images

### ISSUE THREAD

**Issue by eckseleven** [rancher/rke#1795]
**RKE version:** `v0.3.2`

**Docker version: (`docker version`,`docker info` preferred)**  `17.03.2-ce`

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
cat /etc/os-release:
`NAME="Amazon Linux AMI"`
`VERSION="2018.03"`
`ID="amzn"`
`ID_LIKE="rhel fedora"`
`VERSION_ID="2018.03"`
`PRETTY_NAME="Amazon Linux AMI 2018.03"`
`ANSI_COLOR="0;33"`
`CPE_NAME="cpe:/o:amazon:linux:2018.03:ga"`
`HOME_URL="http://aws.amazon.com/amazon-linux-ami/"`

uname -r:
`4.14.138-89.102.amzn1.x86_64`

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)** `AWS EC2`

**cluster.yml file:**
`ssh_key_path: /home/ec2-user/.ssh/id_rsa`
`nodes:`
`- address: <IP1>`
`  internal_address: <IP1>`
`  user: ec2-user`
`  role: [ "controlplane", "etcd", "worker" ]`
`- address: <IP2>`
`  internal_address: <IP2>`
`  user: ec2-user`
`  role: [ "controlplane", "etcd", "worker" ]`
`- address: <IP3>`
`  internal_address: <IP3>`
`  user: ec2-user`
`  role: [ "controlplane", "etcd", "worker" ]`
` `
`private_registries:`
`- url: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo`
`  is_default: true`
` `
`services:`
`  kubelet:`
`    extra_binds:`
`    - "/cgroup:/sys/fs/cgroup"`
` `
`cloud_provider:`
`  name: aws`
` `
`system_images:`
`  etcd: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_coreos-etcd_v3.3.10-rancher1`
`  kubernetes: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_hyperkube_v1.15.5-rancher1`
`  alpine: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_rke-tools_v0.1.50`
`  nginx_proxy: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_rke-tools_v0.1.50`
`  cert_downloader: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_rke-tools_v0.1.50`
`  kubernetes_services_sidecar: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_rke-tools_v0.1.50`
`  kube_dns: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_k8s-dns-kube-dns_1.15.0`
`  dnsmasq: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_k8s-dns-dnsmasq-nanny_1.15.0`
`  kube_dns_sidecar: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_k8s-dns-sidecar_1.15.0`
`  flannel: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_coreos-flannel_v0.11.0-rancher1`
`  flannel_cni: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_flannel-cni_v0.3.0-rancher5`
`  calico_node: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_calico-node_v3.7.4`
`  calico_cni: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_calico-cni_v3.7.4`
`  calico_ctl: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_calico-ctl_v2.0.0`
`  calico_controllers: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_calico-kube-controllers_v3.7.4`
`  canal_node: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_calico-node_v3.7.4`
`  canal_cni: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_calico-cni_v3.7.4`
`  canal_flannel: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_coreos-flannel_v0.11.0`
`  pod_infra_container: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_pause_3.1`
`  ingress: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_nginx-ingress-controller_nginx-0.25.1-rancher1`
`  ingress_backend: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_nginx-ingress-controller-defaultbackend_1.5-rancher1`
`  metrics_server: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_metrics-server_v0.3.3`
`  core_dns: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_coredns-coredns_1.3.1`
`  core_dns_autoscaler: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_cluster-proportional-autoscaler_1.3.0`

**Steps to Reproduce:**
- Push all required images to to ECR repo
- Run `rke up` with cluster configuration above. 

**Results:**
Cluster is not created. Output from RKE indicates 'docker login' is not being performed. The instances are configured with an instance role that permits pull access to the ECR repo.

Output contains messages similar to:
`WARN[0003] Failed to create Docker container [cluster-state-deployer] on host [<IP1>]: Error response from daemon: No such image: 111222333444.dkr.ecr.us-gov-west-1.amazonaws.com/my-repo:rancher_rke-tools_v0.1.50`

I understand that when it comes time to run my actual workloads ECR is supported, but is that not the case for Rancher/K8s infrastructure images?

**Comment by eckseleven** [comment]
An update to this. I've changed my private registry section of the cluster yaml to look like this with information obtained from 'aws ecr get-login'.

`private_registries:`
`- url: 111.222.333.444.dkr.ecr.us-gov-west-1.amazonaws.com`
`  user: AWS`
`  password: XXXXXXXXXXX`
`  is_default: true`

Things do get further; I see images getting pulled now. But RKE eventually fails abruptly with a stack trace:

`INFO[0052] Kubernetes images pulled successfully`
`WARN[0052] rancher_coreos is not in dotted-tri format`
`panic: runtime error: invalid memory address or nil pointer dereference [recovered]`
`        panic: runtime error: invalid memory address or nil pointer dereference`
`[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x137c686]`
` `
`goroutine 1 [running]:`
`github.com/rancher/rke/vendor/github.com/urfave/cli.HandleAction.func1(0xc00060ad28)`
`        /go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/app.go:472 +0x287`
`panic(0x1510320, 0x28848a0)`
`        /usr/local/go/src/runtime/panic.go:522 +0x1b5`
`github.com/rancher/rke/cluster.(*Cluster).BuildEtcdProcess(0xc0000ba400, 0xc0003d2600, 0x0, 0x0, 0x0, 0x28398ef, 0x1, 0x0, 0x0, 0x0, ...)`
`        /go/src/github.com/rancher/rke/cluster/plan.go:958 +0x1bd6`
`github.com/rancher/rke/cluster.BuildRKEConfigNodePlan(0x1a14e00, 0xc000040080, 0xc0000ba400, 0xc0003d2600, 0xc0002566c0, 0x3b, 0x1, 0x0,` `0x0, 0x1, ...)`
`        /go/src/github.com/rancher/rke/cluster/plan.go:110 +0x7ae`
`github.com/rancher/rke/cluster.(*Cluster).DeployControlPlane(0xc0000ba400, 0x1a14e00, 0xc000040080, 0xc000608fe8, 0x0, 0xc000154840)`
`        /go/src/github.com/rancher/rke/cluster/cluster.go:97 +0x195`
`github.com/rancher/rke/cmd.ClusterUp(0x1a14e00, 0xc000040080, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7ffd6b934685, 0x13, 0x0, ...)`
`        /go/src/github.com/rancher/rke/cmd/up.go:148 +0x870`
`github.com/rancher/rke/cmd.clusterUpFromCli(0xc000178b40, 0x0, 0x0)`
`        /go/src/github.com/rancher/rke/cmd/up.go:245 +0x71a`
`reflect.Value.call(0x14bc4c0, 0x179fe38, 0x13, 0x16c8fd3, 0x4, 0xc0004bccc8, 0x1, 0x1, 0xc00003a2c0, 0x4120f4, ...)`
`        /usr/local/go/src/reflect/value.go:447 +0x461`
`reflect.Value.Call(0x14bc4c0, 0x179fe38, 0x13, 0xc0004bccc8, 0x1, 0x1, 0xc000062e08, 0x140, 0xc000062e00)`
`        /usr/local/go/src/reflect/value.go:308 +0xa4`
`github.com/rancher/rke/vendor/github.com/urfave/cli.HandleAction(0x14bc4c0, 0x179fe38, 0xc000178b40, 0x0, 0x0)`
`        /go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/app.go:481 +0x1ff`
`github.com/rancher/rke/vendor/github.com/urfave/cli.Command.Run(0x16c8b37, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x16d7f2a, 0x14, 0x0, ...)`
`        /go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/command.go:186 +0x8d1`
`github.com/rancher/rke/vendor/github.com/urfave/cli.(*App).Run(0xc000112900, 0xc00003a080, 0x4, 0x4, 0x0, 0x0)`
`        /go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/app.go:235 +0x601`
`main.mainErr(0xc00007e540, 0x19dd5e0)`
`        /go/src/github.com/rancher/rke/main.go:66 +0x10c8`
`main.main()`
`        /go/src/github.com/rancher/rke/main.go:22 +0x50`

Any help is appreciated.

**Comment by eckseleven** [comment]
I've retagged the etcd image from **rancher_coreos-etcd_v3.3.10-rancher1** to **v3.3.10-coreos-etcd-rancher1** in my ECR repo and now I'm past the last issue. Now I hit the next issue. Here is the RKE output (debug on):

`INFO[0134] [network] Setting up network plugin: canal`
`INFO[0134] [addons] Saving ConfigMap for addon rke-network-plugin to Kubernetes`
`INFO[0134] [addons] Successfully saved ConfigMap for addon rke-network-plugin to Kubernetes`
`INFO[0134] [addons] Executing deploy job rke-network-plugin`
`DEBU[0135] [k8s] waiting for job rke-network-plugin-deploy-job to complete..`
`FATA[0165] Failed to get job complete status for job rke-network-plugin-deploy-job in namespace kube-system`

If I hop on one of the nodes, I see the rke-network-plugin-depoy-job running and crashing repeatedly. If I check the logs for one of the failed containers I see this:

`clusterrole.rbac.authorization.k8s.io/calico unchanged`
`clusterrolebinding.rbac.authorization.k8s.io/calico-node unchanged`
`clusterrole.rbac.authorization.k8s.io/flannel unchanged`
`clusterrolebinding.rbac.authorization.k8s.io/canal-flannel unchanged`
`clusterrolebinding.rbac.authorization.k8s.io/canal-calico unchanged`
`configmap/canal-config unchanged`
`serviceaccount/canal unchanged`
`customresourcedefinition.apiextensions.k8s.io/felixconfigurations.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/bgpconfigurations.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/ippools.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/hostendpoints.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/clusterinformations.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/globalnetworkpolicies.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/globalnetworksets.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/networkpolicies.crd.projectcalico.org unchanged`
`customresourcedefinition.apiextensions.k8s.io/networksets.crd.projectcalico.org unchanged`
`The DaemonSet "canal" is invalid:`
`* spec.template.spec.containers[0].securityContext.privileged: Forbidden: disallowed by cluster policy`
`* spec.template.spec.containers[1].securityContext.privileged: Forbidden: disallowed by cluster policy`

And `docker inspect` for the failed container reveals it is running `kubectl apply -f /etc/config/rke-network-plugin.yaml`. I've been unable to find what this yaml contains but considering the struggles I've had using ECR in the past, my guess is an image name and tag in there doesn't align with what I have in ECR.

If I push all the necessary images to an ordinary Docker private registry (registry:2 image) and use the same yaml above with just the private registry and image tags changed, it all works fine. We are hoping to use ECR however, so we don't need to managed a private registry.

**Comment by eckseleven** [comment]
I found a way to get this working. I ended up changing how my ECR repos are set up and how the images are organized in them. I was hoping to keep all infrastructure images in a single repo where the image tag would contain the image name plus version, something like "rancher_hyperkube_v1.15.5-rancher1" (see above in the cluster yaml). But despite listing out all the images in the cluster yaml, RKE must not use them because things wouldn't fully start.

After I added several new ECR repos (for example, a repo named "rancher/hypercube") and then tagged images with just the version, I was able to get everything to start (obviously with an updated cluster yaml).

At this point I'm not sure this is a real issue anymore. It seems a little odd that RKE doesn't use the system images I tell it to, however I understand I was tagging my images in an unconventional way.
